src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagKey.cs

```diff
@@ -0,0 +1,428 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Mvc.TagHelpers
+{
+    /// <summary>
+    /// <see cref="TagHelper"/> base implementation for caching elements.
+    /// </summary>
+    public class CacheTagKey : IEquatable<CacheTagKey>
+    {
+        private const string CacheKeyTokenSeparator = "||";
+        private static readonly char[] AttributeSeparator = new[] { ',' };
+
+        private string _key;
+        private string _prefix;
+        private string _varyBy;
+        private DateTimeOffset? _expiresOn;
+        private TimeSpan? _expiresAfter;
+        private TimeSpan? _expiresSliding;
+        private IList<KeyValuePair<string, string>> _headers;
+        private IList<KeyValuePair<string, string>> _queries;
+        private IList<KeyValuePair<string, string>> _routeValues;
+        private IList<KeyValuePair<string, string>> _cookies;
+        private bool _varyByUser;
+        private string _username;
+
+        private string _generatedKey;
+        private int? _hashcode;
+        
+        private CacheTagKey()
+        {
+        }
+
+        public static CacheTagKey From(CacheTagHelper tagHelper, TagHelperContext context)
```

Comment:
`From(CacheTagHelperBase tagHelper, TagHelperContext context)` and `From(DistributedCacheTagHelper tagHelper, TagHelperContext context)` should enable you to reuse a lot of what's done in this method.